### DISCUSSION TITLE
WinDbg kernel debugging steps

### DISCUSSION THREAD

**Discussion by abcde-r** [NationalSecurityAgency/ghidra#5189] [Q&A] [answered]
I am using Ghidra 10.2.3. I currently can use WinDbg (classic, not Preview) to connect from the debugger VM to the debuggee VM via network-based debugging. But I'd like to do this from within Ghidra, so that I can see the code in the decompiler while I step through the assembly. I am running Ghidra on the Windows-based debugger VM.

I've looked at the other threads, but the [newest](https://github.com/NationalSecurityAgency/ghidra/discussions/3034) one is almost 2 years old, so I'm not sure how relevant it still is (e.g. I don't see any "Connect" UI as shown at the very beginning.)

I start by clicking on the Debugger icon in the toolchest

![image](https://user-images.githubusercontent.com/127523449/229604965-d6cdadbb-2825-417f-822c-4cc120324135.png)

From there, basically everything's greyed out, and I don't see a way to just connect via WinDbg to the remote system.

![image](https://user-images.githubusercontent.com/127523449/229605207-f3529a5c-0860-4639-afe7-9b281fc5a312.png)

If I select File -> Open and choose the ntoskrnl.exe which I've already imported into my project, then some debugger configuration options become available:

![image](https://user-images.githubusercontent.com/127523449/229605549-f06540b6-c800-47fd-922d-cbb31605af26.png)

If I select the first one "Debug ntoskrnl.exe in dbgeng locally via GADP", then I get an error in the interpreter window saying "The %1 application cannot be run in Win32 mode.". Which makes sense, because of course the kernel can't be run that way, and that's not the way I'm wanting to run it.

**So I'm at a loss for how to continue from here.** Where is the UI to remotely connect to the other VM that happens to have a copy of this ntoskrnl.exe?

**Comment by d-millar** [comment]
@abcde-r Hi and, yes, not perhaps the most obvious thing....  

What you want to do is the following:
(1)  Leave ntoskrnl.exe in the Static Listing.  This is unnecessary for kernel debugging, but it will allow you to associate that program with the Dynamic view of memory in later steps.
(2) In the Debugger Targets window, select "IN-VM MS dbgeng local debugger".  (You can also use the GADP variant, but, in this case, that buys you very little.)
(3) For the DebugConnect options, enter the options you would normally pass to Windbg when you connect to a kernel target, e.g. "net:port=12345,key=1.1.1.1" (depending, of course, on the options you specified using bcdedit on the target).
(4) Hit "Connect".

I should mention that a large number of improvements have been made recently specifically for Windows kernel debugging, so, if you're feeling brave, I would definitely consider building "master" from the source and using that.

Let us know if any of the above doesn't work or you get stuck elsewhere!
D

**Reply by abcde-r** [reply]
Thanks for such a fast reply!

I select the "in dbgeng locally IN-VM" as shown:
![image](https://user-images.githubusercontent.com/127523449/229629368-e1c2eff5-1fc0-4b90-96cc-462f58ba7869.png)

And then I set the following options. But you can see that there's the "Non-respawnable executor terminated unexpectedly" error that shows up on the side, and it never succeeds in connecting.
![image](https://user-images.githubusercontent.com/127523449/229629213-319861ec-e4a5-4db6-8c67-da7566eaf7a5.png)

You said "depending on the options you specified using bcdedit on the target". In bcdedit one specifies the IP address of the debugger VM. So I tried with the debugger VM IP. But when that didn't work I tried the debuggee VM IP, and no server= option at all (since one doesn't give that in WinDbg when starting the kernel debugging session.) However, none of these yielded a successful connection.

I also had tried following the guidance from [this thread](https://github.com/NationalSecurityAgency/ghidra/discussions/4949) and breaking in to WinDbg and calling .server:
```
0: kd> .server tcp:port=54321
Server started.  Client can connect with any of these command lines
0: <debugger> -remote tcp:Port=54321,Server=hostname
```
And then tried connecting with only the "net:port=54321" option, but it yielded the same error.

**Comment by d-millar** [comment]
I feel like "ip" would be the correct key, but shouldn't be necessary (?).  Testing this right now.

**Reply by d-millar** [reply]
Hmmm, am having some issues getting my test set-up running - apologies - but, in the meantime, a quick question.  You can connect from windbg to the target, yes?  If so, the same options for the kernel connection should work.  Also, judging from your snapshot, you're still connecting from the ToolBar.  While that might work, I'm not positive it will, which is why I suggested using the Debugger Targets window.  Once my test harness is back to working, will test that however.

**Comment by d-millar** [comment]
OOOPS - I totally messed up and have led you down the completely wrong path.  My sincerest apologies!  Try this:

(1) From "Debugger Targets" launch "IN-VM MS dbgeng local debugger, as above.  **(NOT from ToolBar!)**
(2) Open the Connectors node in the Objects view.
(3) Select "Attach to kernel" node.
(4) While it's selected, hit the "Launch (X)" button in Objects.
(5) In the Options dialog, enter "net:port=12345, key=1.2.3.4" or equivalent and hit "Launch".
(6) Cross your fingers and hope I didn't give your more bad advice.

FYI, the instructions I gave you before are for connecting through windbg, i.e. if you had connected Windbg to the target and then issued a .server command.  Too many options for my Monday brain apparently.

**Reply by abcde-r** [reply]
OK, so the ask is to launch it from the Debugger Targets window, like this, correct?

![image](https://user-images.githubusercontent.com/127523449/229655463-aa34d777-ed1c-4237-abe9-cbf3dca2ac3b.png)

That doesn't seem to yield any difference (other than the fact that when done that way rather than through the menu bar, the pulldown menu is pull-downable in a way that it isn't otherwise.)

![image](https://user-images.githubusercontent.com/127523449/229655604-10ca8df2-0a69-4cda-b0d8-953c3dcf720a.png)

However, when I select "IN-VM MS dbgeng local debugger", I try the below unsuccessfully:
"tcp:Port=54321,server=172.16.126.137" (for local connection to WinDbg where I ran .server tcp:Port=54321)
![image](https://user-images.githubusercontent.com/127523449/229658263-b0493bda-0c3a-4bf9-b982-2939cd8e76db.png)

And I tried various permutations like the below for an attempted direct connection:
"tcp:Port=50505,key=a.b.c.d{,{server,ip}=172.16.126.13{7,8}}"

I've never tried to use the CLI invocation of WinDbg before, but what I can confirm is that if I do
".server tcp:port=54321"
in an already kernel-debugging WinDbg instance, followed by:
"C:\Program Files (x86)\Windows Kits\10\Debuggers\x64>windbg -remote tcp:Port=54321,Server=172.16.126.137"
**does** pop up a 2nd WinDbg instance, that appears to be connected due to the first one (they both break at the same time, list the same output, etc.)

But again, that same "tcp:Port=54321,Server=172.16.126.137" when passed in the "IN-VM MS dbgeng local debugger" UI's "DebugConnect options (.server)" text box, doesn't yield a connection. It just yields a "Non-respawnable executor terminated unexpectedly" error and indefinite "Connecting..." prompt.

---

And FWIW if I try the GADP option instead:

![image](https://user-images.githubusercontent.com/127523449/229658762-d067b32a-9f2e-42eb-b190-5bd0e2fe6a6d.png)

I get some different (but maybe related?) Java error:

![image](https://user-images.githubusercontent.com/127523449/229658821-c48169df-d982-451b-9c83-6ad3702b1b80.png)

**Reply by d-millar** [reply]
Just to clarify, when you connect to the IN-VM dbgeng target, do NOT put anything in the parameters there.  Both should say "none".

**Reply by d-millar** [reply]
I.e. you're adding the "net:port..." parameters only to the "Launch (X)" dialog.

**Reply by abcde-r** [reply]
OK, thanks for the clarification that ```Both should say "none".```. I had tried deleting the .server options too, and that yielded the same error. But putting ```"none"``` in there made it connect!

![image](https://user-images.githubusercontent.com/127523449/229666013-a5a56273-5bed-4538-a6ba-20cd8dcfd1ec.png)

Then I hit the "Launch (X)" button
![image](https://user-images.githubusercontent.com/127523449/229666033-ac99a17b-7073-462e-992e-91fbb9b32ebb.png)

And enter "net:port=50505,key=a.b.c.d", but I encounter an error:
![image](https://user-images.githubusercontent.com/127523449/229666465-e8563fb9-da90-461c-bc85-5dfd1ea0f6a4.png)
![image](https://user-images.githubusercontent.com/127523449/229667593-e44e420f-8809-42eb-aace-e4d8bd5c7a6c.png)

But hey, that's progress! Thanks again for getting me closer! Ideas for next-stage debugging the debugger?

---
On the other hand, if I keep trying the local WinDbg ".server tcp:port=54321" option, and then instead put "net:port=54321", I get no error and an encouraging "Waiting to reconnect..."

![image](https://user-images.githubusercontent.com/127523449/229668178-22489ac7-03b4-4ce5-8fe4-2269238be24b.png)

However, the local WinDbg doesn't seem to think anything has connected to it, and I can't seem to find a way to break from within Ghidra in a way that stops the remote kernel target.

**Reply by d-millar** [reply]
Hmmmm, well, I should definitely fix that - blank really should behave the same as "none".  

That said, back to the problem at hand, I would like to know if windbg connects to the target via the normal kernel connection, i.e. start with "kernel debug", pick the Net tab, enter 50505 for the Port and a.b.c.d for the Key.  If that doesn't connect, something else is going on.  (You shouldn't need to use ".server" or "remote" really - those are for relaying connections.)  

If that fails, I might use "netstat -a" to see if the port is tied up for some reason. Could be one of our previous experiments has ghosted a process and is hogging the port.  (Alternatively, I suppose you could pick a different port, but that seems like a hassle.)

**Reply by d-millar** [reply]
not typing too well - edits on the github page

**Reply by abcde-r** [reply]
Yes, the normal WinDbg GUI-based connection is still working fine.

**Comment by d-millar** [comment, marked as answer]
OK, I think I'm not understanding your set-up correctly, so am going to re-state some things which are probably already obvious to you in the hopes of clariyfing my understanding.  Apologies in advance.

So, I'm assuming we have ony two machines in play (possibly viirtual).  For the sake of argument, let's say the machine with the debugger lives at 192.168.1.1, and the target machine lives at 192.168.1.2.  You've used bdcedit on the latter to set "debug" on and "dbgsettings" to "net:port=50505,key=a.b.c.d,server=192.168.1.1".  Running Windbg (x64) from 192.168.1.1 using "Kernel Debug..." with "Port Number"=50505 and "Key"=a.b.c.d from the "NET" tab connects correctly and allows you to break and do stuff.

There should be (at least) two possible scenarios for connecting Ghidra to the target:  one in which you leave Windbg running and one in which you terminate Windbg and rely on the Ghidra JNA-wrapped dbgeng loaded into its JVM.

In the former case (which I would not recommend),  you connect with Windbg, break, and issue something like ".server tcp:port=54321".  Then, in Ghidra, you connect from "Debugger Targets" choosing IN-VM dbgeng with "DebugConnect options (.server)" set to "tcp:54321".

In the latter case (which I would say is preferable), you make sure Windbg is NOT running and has freed the port.  In Ghidra, you connect from "Debugger Targets" choosing IN-VM dbgeng with "none"/"none".  Once connected, you select Objects->Connectors->Attach to kernel and specify "net:port=50505,key=a.b.c.d" in "Options" in the dialog launched by "Launch (X)".

It looks like you did that, but then got the 0x80004005 error, which suggests the port was either unavailable or either the port or key were incorrect.  Does that seem accurate?  One thing that might help a little is to connect with Windbg, break, disconnect leaving the target paused, and then reconnect with Ghidra.  That tends to make the state a little more obvious, although you should be able to connect with either the target running or stopped.  If running, you should be able to break using either Pause button in Objects or the Interpreter or by hitting Ctrl-I from the Interpreter (Ctrl-C is already used for copy).

**Reply by abcde-r** [reply]
OK, I think the issue was that *I didn't realize I needed to close the existing connected WinDbg in order to connect with Ghidra!*

In the process of debugging I also changed out the port to 55555, key to a.a.a.a, and and restarted the debugger and debuggee VM. I confirmed I could connect with WinDbg, and then I closed it. I then confirmed with netstat that I didn't see any connections on port 55555. Once I did that, and went through the steps as prescribed before, **I was able to connect successfully!** I doubt the 55555 vs. 50505 matters, and it's probably all down to still having WinDbg connected. Thanks!

**Comment by d-millar** [comment]
Excellent - glad to hear you got things going.  Sorry it took so long!  (And thanks re typo - edited for the next customer.)
